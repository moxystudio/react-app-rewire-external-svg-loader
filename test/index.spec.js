'use strict';

const serializerPath = require('jest-serializer-path');
const cloneDeep = require('lodash.clonedeep');
const rewireExternalSvgLoader = require('..');

expect.addSnapshotSerializer(serializerPath);

const mockConfig = {
    module: {
        rules: [
            {
                test: /\.(js|jsx|mjs)$/,
                enforce: 'pre',
                use: [
                    { options: {}, loader: 'path/to/eslint-loader/index.js' },
                ],
                include: 'path/to/src',
            },
            {
                oneOf: [
                    {
                        test: [/\.bmp$/, /\.gif$/, /\.jpe?g$/, /\.png$/],
                        loader: 'path/to/url-loader/index.js',
                        options: {},
                    },
                    {
                        test: /\.(js|jsx|mjs)$/,
                        include: 'path/to/src',
                        loader: 'path/to/babel-loader/lib/index.js',
                        options: {},
                    },
                    {
                        test: /\.css$/,
                        use: [
                            'path/to/style-loader/index.js',
                            {
                                loader: 'path/to/css-loader/index.js',
                                options: { importLoaders: 1 },
                            },
                            {
                                loader: 'path/to/postcss-loader/lib/index.js',
                                options: {},
                            },
                        ],
                    },
                    {
                        exclude: [/\.js$/, /\.html$/, /\.json$/],
                        loader: 'path/to/file-loader/dist/cjs.js',
                        options: { name: 'static/media/[name].[hash:8].[ext]' },
                    },
                ],
            },
        ],
    },
    plugins: [],
};

it('should modify the webpack config correctly for development', () => {
    const result = rewireExternalSvgLoader(cloneDeep(mockConfig), 'development');

    expect(result).toMatchSnapshot();
});

it('should modify the webpack config correctly for production', () => {
    const result = rewireExternalSvgLoader(cloneDeep(mockConfig), 'production');

    expect(result).toMatchSnapshot();
});

it('should override the default test / include / exclude', () => {
    const result = rewireExternalSvgLoader(cloneDeep(mockConfig), 'development', {
        test: 'svg',
        include: 'foo',
        exclude: 'bar',
    });

    expect(result).toMatchSnapshot();
});

it('should pass options to the loader', () => {
    const result = rewireExternalSvgLoader(cloneDeep(mockConfig), 'development', {
        loaderOptions: { name: 'foo.svg' },
    });

    expect(result).toMatchSnapshot();
});

it('should pass options to the plugin', () => {
    const result = rewireExternalSvgLoader(cloneDeep(mockConfig), 'development', {
        pluginOptions: { emit: false },
    });

    expect(result).toMatchSnapshot();
});

it('should allow usage with compose', () => {
    expect(rewireExternalSvgLoader(cloneDeep(mockConfig), 'development'))
    .toEqual(rewireExternalSvgLoader()(cloneDeep(mockConfig), 'development'));

    expect(rewireExternalSvgLoader(cloneDeep(mockConfig), 'development', { include: 'foo' }))
    .toEqual(rewireExternalSvgLoader({ include: 'foo' })(cloneDeep(mockConfig), 'development'));
});
