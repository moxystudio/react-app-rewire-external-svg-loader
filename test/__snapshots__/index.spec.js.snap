// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should modify the webpack config correctly for development 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "enforce": "pre",
        "include": "path/to/src",
        "test": /\\\\\\.\\(js\\|jsx\\|mjs\\)\\$/,
        "use": Array [
          Object {
            "loader": "path/to/eslint-loader/index.js",
            "options": Object {},
          },
        ],
      },
      Object {
        "oneOf": Array [
          Object {
            "loader": "path/to/url-loader/index.js",
            "options": Object {},
            "test": Array [
              /\\\\\\.bmp\\$/,
              /\\\\\\.gif\\$/,
              /\\\\\\.jpe\\?g\\$/,
              /\\\\\\.png\\$/,
            ],
          },
          Object {
            "include": "path/to/src",
            "loader": "path/to/babel-loader/lib/index.js",
            "options": Object {},
            "test": /\\\\\\.\\(js\\|jsx\\|mjs\\)\\$/,
          },
          Object {
            "test": /\\\\\\.css\\$/,
            "use": Array [
              "path/to/style-loader/index.js",
              Object {
                "loader": "path/to/css-loader/index.js",
                "options": Object {
                  "importLoaders": 1,
                },
              },
              Object {
                "loader": "path/to/postcss-loader/lib/index.js",
                "options": Object {},
              },
            ],
          },
          Object {
            "exclude": undefined,
            "include": "<PROJECT_ROOT>/src",
            "test": /\\\\\\.svg\\$/,
            "use": Array [
              Object {
                "loader": "<PROJECT_ROOT>/node_modules/external-svg-sprite-loader/index.js",
                "options": Object {
                  "name": "static/media/svg-sprite.svg",
                },
              },
              Object {
                "loader": "<PROJECT_ROOT>/node_modules/svg-css-modules-loader/src/svg-css-modules-loader.js",
                "options": Object {
                  "transformId": true,
                },
              },
            ],
          },
          Object {
            "exclude": Array [
              /\\\\\\.js\\$/,
              /\\\\\\.html\\$/,
              /\\\\\\.json\\$/,
            ],
            "loader": "path/to/file-loader/dist/cjs.js",
            "options": Object {
              "name": "static/media/[name].[hash:8].[ext]",
            },
          },
        ],
      },
    ],
  },
  "plugins": Array [
    SvgStorePlugin {
      Symbol(): true,
      Symbol(): Object {},
    },
  ],
}
`;

exports[`should modify the webpack config correctly for production 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "enforce": "pre",
        "include": "path/to/src",
        "test": /\\\\\\.\\(js\\|jsx\\|mjs\\)\\$/,
        "use": Array [
          Object {
            "loader": "path/to/eslint-loader/index.js",
            "options": Object {},
          },
        ],
      },
      Object {
        "oneOf": Array [
          Object {
            "loader": "path/to/url-loader/index.js",
            "options": Object {},
            "test": Array [
              /\\\\\\.bmp\\$/,
              /\\\\\\.gif\\$/,
              /\\\\\\.jpe\\?g\\$/,
              /\\\\\\.png\\$/,
            ],
          },
          Object {
            "include": "path/to/src",
            "loader": "path/to/babel-loader/lib/index.js",
            "options": Object {},
            "test": /\\\\\\.\\(js\\|jsx\\|mjs\\)\\$/,
          },
          Object {
            "test": /\\\\\\.css\\$/,
            "use": Array [
              "path/to/style-loader/index.js",
              Object {
                "loader": "path/to/css-loader/index.js",
                "options": Object {
                  "importLoaders": 1,
                },
              },
              Object {
                "loader": "path/to/postcss-loader/lib/index.js",
                "options": Object {},
              },
            ],
          },
          Object {
            "exclude": undefined,
            "include": "<PROJECT_ROOT>/src",
            "test": /\\\\\\.svg\\$/,
            "use": Array [
              Object {
                "loader": "<PROJECT_ROOT>/node_modules/external-svg-sprite-loader/index.js",
                "options": Object {
                  "name": "static/media/svg-sprite.[hash:8].svg",
                },
              },
              Object {
                "loader": "<PROJECT_ROOT>/node_modules/svg-css-modules-loader/src/svg-css-modules-loader.js",
                "options": Object {
                  "transformId": true,
                },
              },
            ],
          },
          Object {
            "exclude": Array [
              /\\\\\\.js\\$/,
              /\\\\\\.html\\$/,
              /\\\\\\.json\\$/,
            ],
            "loader": "path/to/file-loader/dist/cjs.js",
            "options": Object {
              "name": "static/media/[name].[hash:8].[ext]",
            },
          },
        ],
      },
    ],
  },
  "plugins": Array [
    SvgStorePlugin {
      Symbol(): true,
      Symbol(): Object {},
    },
  ],
}
`;

exports[`should override the default test / include / exclude 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "enforce": "pre",
        "include": "path/to/src",
        "test": /\\\\\\.\\(js\\|jsx\\|mjs\\)\\$/,
        "use": Array [
          Object {
            "loader": "path/to/eslint-loader/index.js",
            "options": Object {},
          },
        ],
      },
      Object {
        "oneOf": Array [
          Object {
            "loader": "path/to/url-loader/index.js",
            "options": Object {},
            "test": Array [
              /\\\\\\.bmp\\$/,
              /\\\\\\.gif\\$/,
              /\\\\\\.jpe\\?g\\$/,
              /\\\\\\.png\\$/,
            ],
          },
          Object {
            "include": "path/to/src",
            "loader": "path/to/babel-loader/lib/index.js",
            "options": Object {},
            "test": /\\\\\\.\\(js\\|jsx\\|mjs\\)\\$/,
          },
          Object {
            "test": /\\\\\\.css\\$/,
            "use": Array [
              "path/to/style-loader/index.js",
              Object {
                "loader": "path/to/css-loader/index.js",
                "options": Object {
                  "importLoaders": 1,
                },
              },
              Object {
                "loader": "path/to/postcss-loader/lib/index.js",
                "options": Object {},
              },
            ],
          },
          Object {
            "exclude": "bar",
            "include": "foo",
            "test": "svg",
            "use": Array [
              Object {
                "loader": "<PROJECT_ROOT>/node_modules/external-svg-sprite-loader/index.js",
                "options": Object {
                  "name": "static/media/svg-sprite.svg",
                },
              },
              Object {
                "loader": "<PROJECT_ROOT>/node_modules/svg-css-modules-loader/src/svg-css-modules-loader.js",
                "options": Object {
                  "transformId": true,
                },
              },
            ],
          },
          Object {
            "exclude": Array [
              /\\\\\\.js\\$/,
              /\\\\\\.html\\$/,
              /\\\\\\.json\\$/,
            ],
            "loader": "path/to/file-loader/dist/cjs.js",
            "options": Object {
              "name": "static/media/[name].[hash:8].[ext]",
            },
          },
        ],
      },
    ],
  },
  "plugins": Array [
    SvgStorePlugin {
      Symbol(): true,
      Symbol(): Object {},
    },
  ],
}
`;

exports[`should pass options to the loader 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "enforce": "pre",
        "include": "path/to/src",
        "test": /\\\\\\.\\(js\\|jsx\\|mjs\\)\\$/,
        "use": Array [
          Object {
            "loader": "path/to/eslint-loader/index.js",
            "options": Object {},
          },
        ],
      },
      Object {
        "oneOf": Array [
          Object {
            "loader": "path/to/url-loader/index.js",
            "options": Object {},
            "test": Array [
              /\\\\\\.bmp\\$/,
              /\\\\\\.gif\\$/,
              /\\\\\\.jpe\\?g\\$/,
              /\\\\\\.png\\$/,
            ],
          },
          Object {
            "include": "path/to/src",
            "loader": "path/to/babel-loader/lib/index.js",
            "options": Object {},
            "test": /\\\\\\.\\(js\\|jsx\\|mjs\\)\\$/,
          },
          Object {
            "test": /\\\\\\.css\\$/,
            "use": Array [
              "path/to/style-loader/index.js",
              Object {
                "loader": "path/to/css-loader/index.js",
                "options": Object {
                  "importLoaders": 1,
                },
              },
              Object {
                "loader": "path/to/postcss-loader/lib/index.js",
                "options": Object {},
              },
            ],
          },
          Object {
            "exclude": undefined,
            "include": "<PROJECT_ROOT>/src",
            "test": /\\\\\\.svg\\$/,
            "use": Array [
              Object {
                "loader": "<PROJECT_ROOT>/node_modules/external-svg-sprite-loader/index.js",
                "options": Object {
                  "name": "foo.svg",
                },
              },
              Object {
                "loader": "<PROJECT_ROOT>/node_modules/svg-css-modules-loader/src/svg-css-modules-loader.js",
                "options": Object {
                  "transformId": true,
                },
              },
            ],
          },
          Object {
            "exclude": Array [
              /\\\\\\.js\\$/,
              /\\\\\\.html\\$/,
              /\\\\\\.json\\$/,
            ],
            "loader": "path/to/file-loader/dist/cjs.js",
            "options": Object {
              "name": "static/media/[name].[hash:8].[ext]",
            },
          },
        ],
      },
    ],
  },
  "plugins": Array [
    SvgStorePlugin {
      Symbol(): true,
      Symbol(): Object {},
    },
  ],
}
`;

exports[`should pass options to the plugin 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "enforce": "pre",
        "include": "path/to/src",
        "test": /\\\\\\.\\(js\\|jsx\\|mjs\\)\\$/,
        "use": Array [
          Object {
            "loader": "path/to/eslint-loader/index.js",
            "options": Object {},
          },
        ],
      },
      Object {
        "oneOf": Array [
          Object {
            "loader": "path/to/url-loader/index.js",
            "options": Object {},
            "test": Array [
              /\\\\\\.bmp\\$/,
              /\\\\\\.gif\\$/,
              /\\\\\\.jpe\\?g\\$/,
              /\\\\\\.png\\$/,
            ],
          },
          Object {
            "include": "path/to/src",
            "loader": "path/to/babel-loader/lib/index.js",
            "options": Object {},
            "test": /\\\\\\.\\(js\\|jsx\\|mjs\\)\\$/,
          },
          Object {
            "test": /\\\\\\.css\\$/,
            "use": Array [
              "path/to/style-loader/index.js",
              Object {
                "loader": "path/to/css-loader/index.js",
                "options": Object {
                  "importLoaders": 1,
                },
              },
              Object {
                "loader": "path/to/postcss-loader/lib/index.js",
                "options": Object {},
              },
            ],
          },
          Object {
            "exclude": undefined,
            "include": "<PROJECT_ROOT>/src",
            "test": /\\\\\\.svg\\$/,
            "use": Array [
              Object {
                "loader": "<PROJECT_ROOT>/node_modules/external-svg-sprite-loader/index.js",
                "options": Object {
                  "name": "static/media/svg-sprite.svg",
                },
              },
              Object {
                "loader": "<PROJECT_ROOT>/node_modules/svg-css-modules-loader/src/svg-css-modules-loader.js",
                "options": Object {
                  "transformId": true,
                },
              },
            ],
          },
          Object {
            "exclude": Array [
              /\\\\\\.js\\$/,
              /\\\\\\.html\\$/,
              /\\\\\\.json\\$/,
            ],
            "loader": "path/to/file-loader/dist/cjs.js",
            "options": Object {
              "name": "static/media/[name].[hash:8].[ext]",
            },
          },
        ],
      },
    ],
  },
  "plugins": Array [
    SvgStorePlugin {
      Symbol(): false,
      Symbol(): Object {},
    },
  ],
}
`;
